### PR TITLE
[RLlib] Removes device infos from state when saving `RModule`s  to checkpoints/states.

### LINKED ISSUE #43905: [RLlib] - `TorchRLModule` cannot be loaded on CPU after training on GPU
### What happened + What you expected to happen

# What happened
I run a cloud experiment with GPUs and trained my algorithm using `framework="torch"` on it. Afterwards I tried to load the checkpoint on my local machine which errored out with the following message: 

```python
Traceback (most recent call last):
  File "/home/simon/git-projects/my-project/tuned_examples/pco/evaluate_best_model_gym.py", line 94, in <module>
    algo = pco.from_checkpoint(checkpoint)
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 357, in from_checkpoint
    return Algorithm.from_state(state)
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 385, in from_state
    new_algo = algorithm_class(config=config)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/util/tracing/tracing_helper.py", line 467, in _resume_span
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 533, in __init__
    super().__init__(
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/tune/trainable/trainable.py", line 161, in __init__
    self.setup(copy.deepcopy(self.config))
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/util/tracing/tracing_helper.py", line 467, in _resume_span
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 754, in setup
    self.learner_group = self.config.build_learner_group(
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm_config.py", line 1091, in build_learner_group
    learner_group = LearnerGroup(config=self, module_spec=rl_module_spec)
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/core/learner/learner_group.py", line 128, in __init__
    self._learner.build()
  File "/home/simon/git-projects/my-project/my-algos/algorithms/pco/pco_learner.py", line 50, in build
    super().build()
  File "/home/simon/git-projects/my-project/.venv-nightly-3-9-12/lib/python3.9/site-packages/ray/rllib/core/learner/torch/torch_learner.py", line 313, in build
    assert self._local_gpu_idx < torch.cuda.device_count(), (
AssertionError: local_gpu_idx 0 is not a valid GPU id or is  not available.
```

# What you expected to happen
That models trained on one machine can be easily deployed on another. Basically, that training and then loading a policy is seamless for users.

### Versions / Dependencies

Trained on Ubuntu 20.0, Python 3.9, Ray nightly, NVIDIA T4

Deployed on Fedora 39, Python 3.9.12, Ray nightly

### Reproduction script

Use `rllib/tuned_examples/cartpole_ppo_envrunner.py` and run it on GPU with `torch`, then load the checkpoint on a CPU-only machine.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

When loading an `RLModule` on CPU from a checkpoint/state that was created from a replica on GPU, an error occurs. This PR fixes this error by forcing the module to save its state in form of `numpy.NdArray`s. 

## Related issue number

Closes #43905 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
I'm not sure I understand the exact need for this additional option to the API. My main argument against this would be: When stuff gets stored to a checkpoint, it should be stored in a device-independent fashion. So the issue at hand here is NOT the loading from the checkpoint, but the saving to the checkpoint beforehand, which - I'm guessing - probably happened in torch.cuda tensors, NOT in numpy format.

Can we rather take the opposite approach to keep the mental model of what a checkpoint should be clean? Always save weights (and other tensor/matrix states) as numpy arrays, never as torch or tf tensors. When loading from a checkpoint, the sequence should be something like:
0) RLModule.from_checkpoint(dir=...)
1) load numpy arrays from `dir` file (pickle?)
2) pack these numpy arrays into a state dict, mapping 
3) Call RLModule.set_state(state_dict=...) -> This automatically converts the numpy contents of `state_dict` into torch tensors (with `self.device` as the device of the already existing RLModule object), then performs a `torch.nn.Module.load_state_dict()` operation using these (cuda?) tensors.

**Comment by simonsays1980** [comment]
> I'm not sure I understand the exact need for this additional option to the API. My main argument against this would be: When stuff gets stored to a checkpoint, it should be stored in a device-independent fashion. So the issue at hand here is NOT the loading from the checkpoint, but the saving to the checkpoint beforehand, which - I'm guessing - probably happened in torch.cuda tensors, NOT in numpy format.
> 
> Can we rather take the opposite approach to keep the mental model of what a checkpoint should be clean? Always save weights (and other tensor/matrix states) as numpy arrays, never as torch or tf tensors. When loading from a checkpoint, the sequence should be something like: 0) RLModule.from_checkpoint(dir=...)
> 
> 1. load numpy arrays from `dir` file (pickle?)
> 2. pack these numpy arrays into a state dict, mapping
> 3. Call RLModule.set_state(state_dict=...) -> This automatically converts the numpy contents of `state_dict` into torch tensors (with `self.device` as the device of the already existing RLModule object), then performs a `torch.nn.Module.load_state_dict()` operation using these (cuda?) tensors.

I agree with your argument that we should ensure that checkpointing is device-independent. This should be the cleanest way of doing this. We should investigate, where exactly this device-dependent checkpointing takes place and fix the problem there. 

I am, however, not so sure, if 3. describes how the workflow runs right now. Here is what makes me wonder: If the `torch.nn.Module.load_state_dict()` would use the `self.device` then it should not matter, if the state contains `numpy.array`s or `torch.tensor` (with their own device attribute) as it uses always the module's device. But no matter what - checkpointing always `numpy.array` will avoid this error anyways.

File: rllib/core/rl_module/rl_module.py [hunk 1]
```diff
@@ -806,11 +809,15 @@ def save_to_checkpoint(self, checkpoint_dir_path: Union[str, pathlib.Path]) -> N
...
-    def from_checkpoint(cls, checkpoint_dir_path: Union[str, pathlib.Path]) -> None:
+    def from_checkpoint(
+        cls,
+        checkpoint_dir_path: Union[str, pathlib.Path],
+    ) -> None:
         """Loads the module from a checkpoint directory.
 
         Args:
             checkpoint_dir_path: The directory to load the checkpoint from.
+            map_location: The device on which the module resides.
```
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:820]: Remove this line?

File: rllib/core/rl_module/marl_module.py [hunk 2]
```diff
@@ -367,6 +367,7 @@ def load_state(
             modules_to_load: The modules whose state is to be loaded from the path. If
                 this is None, all modules that are checkpointed will be loaded into this
                 marl module.
+            map_location: The device the module resides on.
```
Comment (sven1977) [inline rllib/core/rl_module/marl_module.py:370]: Remove this line?

**Review by sven1977 — APPROVED** [review]
Looks good! Thanks for this important fix @simonsays1980 !

Just two nits on the docstrings.

File: rllib/core/rl_module/marl_module.py (hunk 2 above)
Comment (sven1977) [inline rllib/core/rl_module/marl_module.py:370]:
```suggestion
```

File: rllib/core/rl_module/rl_module.py (hunk 1 above)
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:820]:
```suggestion
```

File: rllib/core/rl_module/rl_module.py [hunk 3]
```diff
@@ -806,11 +809,15 @@ def save_to_checkpoint(self, checkpoint_dir_path: Union[str, pathlib.Path]) -> N
         self._save_module_metadata(path, SingleAgentRLModuleSpec)
 
     @classmethod
-    def from_checkpoint(cls, checkpoint_dir_path: Union[str, pathlib.Path]) -> None:
+    def from_checkpoint(
+        cls,
+        checkpoint_dir_path: Union[str, pathlib.Path],
+    ) -> None:
```
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:815]:
```suggestion
    def from_checkpoint(cls, checkpoint_dir_path: Union[str, pathlib.Path]) -> None:
```

File: rllib/core/rl_module/tf/tf_rl_module.py [hunk 4]
```diff
@@ -69,6 +69,9 @@ def save_state(self, dir: Union[str, pathlib.Path]) -> None:
         self.save_weights(path, save_format="tf")
 
     @override(RLModule)
-    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
+    def load_state(
+        self,
+        dir: Union[str, pathlib.Path],
+    ) -> None:
```
Comment (sven1977) [inline rllib/core/rl_module/tf/tf_rl_module.py:75]:
```suggestion
    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
```

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 5]
```diff
@@ -117,10 +118,13 @@ def _module_state_file_name(self) -> pathlib.Path:
...
         path = str(pathlib.Path(dir) / self._module_state_file_name())
-        torch.save(self.state_dict(), path)
+        torch.save(convert_to_numpy(self.state_dict()), path)
 
     @override(RLModule)
-    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
+    def load_state(
+        self,
+        dir: Union[str, pathlib.Path],
+    ) -> None:
```
Comment (sven1977) [inline rllib/core/rl_module/torch/torch_rl_module.py:127]:
```suggestion
    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
```

File: rllib/core/rl_module/rl_module.py [hunk 6]
```diff
@@ -688,7 +688,10 @@ def save_state(self, dir: Union[str, pathlib.Path]) -> None:
         pass
 
     @OverrideToImplementCustomLogic
-    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
+    def load_state(
+        self,
+        dir: Union[str, pathlib.Path],
+    ) -> None:
```
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:694]:
```suggestion
    def load_state(self, dir: Union[str, pathlib.Path]) -> None:
```

**Pull request merged**